### DISCUSSION TITLE
How can I perform the `negative_sampling` by the given nodes?

### DISCUSSION THREAD

**Discussion by songsong0425** [pyg-team/pytorch_geometric#9392] [Q&A] [answered]
Hi, I have a simple question about the `negative_sampling` function.

When I have the example `edge_index` as below, I want to run the `negative_sampling`  based on the nodes(i.e., 0, 1, 2, 4, 5, 8) in the original edges.
```python
edge_index = torch.as_tensor([[0, 0, 1, 2],
                              [8, 1, 5, 4]])
```

But it returned unexpected results including unseen nodes.
```python
negative_sampling(edge_index)
# tensor([[8, 3, 1, 8],
#        [4, 8, 8, 3]])
```

How can I get the random negative edges from the nodes in `edge_index`?

**Comment by rusty1s** [comment]
```
edge_index = torch.as_tensor([[0, 0, 1, 2],
                              [8, 1, 5, 4]])
```
means that you have 9 nodes in total, while node 3 and node 7 are isolated nodes. If you want to exclude them, you have to shrink your `edge_index` down to not include isolated nodes:
```
edge_index = torch.as_tensor([[0, 0, 1, 2],
                              [3, 1, 5, 4]])

**Reply by songsong0425** [reply]
Thank you for your comment. But if the numbers in `edge_index` indicate the node ID, is it impossible to do `negative sampling`? (i.e., in the example, 3,6,7 don't have links) Currently, I made a `dictionary` for the existing nodes and performed `negative sampling` on the dictionary value then remapped to the key. Are there any efficient ways?

**Reply by rusty1s** [reply]
If examples 3, 6, 7 don't have links, then why shouldn't negative sampling sample negative links for these nodes? If you want to completely exclude nodes from negative sampling, it is best to fully discard them as node IDs IMO.

**Reply by songsong0425** [reply]
Sorry for the unclear explanation. Before `negative_sampling`, I performed a **custom ID converting process** across the original dataset, changing the node name to the unique IDs. Therefore, some nodes in the positive sets can be absent in the negative set, and vice versa. In this case, I want to perform sampling on the nodes in negative samples and unique nodes in the positive edges for the transductive setting to supplement fewer negatives. Please tell me if you need more explanation or if my approach is wrong.

**Reply by rusty1s** [reply, marked as answer]
If you have specific constraints on how negatives should look like (e.g., a specific subset of allowed negatives), then I think the most straightforward way is to simply write a custom `negative_sampling` function that fits your needs.

**Reply by songsong0425** [reply]
I see. Thank you for your suggestion. I'll try it.
